ftdc/custom_format.go

```diff
@@ -0,0 +1,549 @@
+	ret := make([]float32, 0, 10*len(mapOrder))
+
+	for _, key := range mapOrder {
+		// Walk over the datum in `mapOrder` to ensure we gather values in the order consistent with
+		// the current schema.
+		stats, exists := datum.Data[key]
+		if !exists {
+			fmt.Println("Missing data. Need to know how much to skip in the output float32")
+			return nil
+		}
+
+		rVal := reflect.ValueOf(stats)
+		if rVal.Kind() == reflect.Pointer {
+			rVal = rVal.Elem()
+		}
+
+		// Use reflection to walk the member fields of an individual set of metric readings. We rely
+		// on reflection always walking fields in the same order.
+		//
+		// Note, while reflection in this way isn't intrinsically expensive, it is making more small
+		// function calls and allocations than some more raw alternatives. For example, we can have
+		// the "schema" keep a (field, offset, type) index and we instead access get a single unsafe
+		// pointer to each structure and walk out index to pull out the relevant numbers.
+		for memberIdx := 0; memberIdx < rVal.NumField(); memberIdx++ {
+			rField := rVal.Field(memberIdx)
+			switch {
+			case rField.CanInt():
+				ret = append(ret, float32(rField.Int()))
+			case rField.CanFloat():
+				ret = append(ret, float32(rField.Float()))
+			default:
+				// Embedded structs? Just grab a global logger for now. A second pass will better
+				// validate inputs/remove limitations. And thread through a proper logger if still
+				// necessary.
+				logging.Global().Warn("Bad number type. Type:", rField.Type())
+				// Ignore via writing a 0 and continue.
+				ret = append(ret, 0)
+			}
+		}
+	}
+
+	return ret
+}
+
+// parse reads the entire contents from `rawReader` and returns a list of `Datum`. If an error
+// occurs, the Datum parsed up to that point will be returned.
+func parse(rawReader io.Reader) ([]Datum, error) {
+	ret := make([]Datum, 0)
+
+	// prevValues are the previous values used for producing the diff bits. This is overwritten when
+	// a new metrics reading is made. and nilled out when the schema changes.
+	var prevValues []float32
+
+	// bufio's Reader allows for peeking and potentially better control over how much data to read
+	// from disk at a time.
+	reader := bufio.NewReader(rawReader)
+	var schema *Schema = nil
+	for {
+		peek, err := reader.Peek(1)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+
+			return ret, err
+		}
+
+		// If the first bit of the first byte is `1`, the next block of data is a schema
+		// document. The rest of the bits (for diffing) are irrelevant and will be zero. Thus the
+		// check against `0x1`.
+		if peek[0] == 0x1 {
+			// Consume the 0x1 byte.
+			_, _ = reader.ReadByte()
+
+			// Read json and position the cursor at the next FTDC document. The JSON reader may
+			// "over-read", so `readSchema` assembles a new reader positioned at the right spot. The
+			// schema bytes themselves are expected to be a list of strings, e.g: `["metricNameOne",
+			// "metricName2"]`.
```

Comment:
It looks like we expect more of the schema columns than just `a list of strings`... it looks like they MUST be in the format ["metricName.MetricField"] and also that all fields of a given name must be contiguous in the array. 

If we are going to keep that design decision, can we put that requirement higher up in the explanation?

Reply:
You're right that the parser makes some assumptions here (which are touched in the write-path).

> If we are going to keep that design decision, can we put that requirement higher up in the explanation?

Not sure how much higher up you're thinking of here.

We should talk offline about options here. I'll be pedantic and say neither of those things need be true to write a parser. I don't think those things don't actually help in writing a parser, but rather are simplifications for writing the "writer" of ftdc data.